### PR TITLE
Fix expression defining entire query in annotated repository methods

### PR DESCRIPTION
**Pull request by rolag-it**
This fix enables defining an entire JSON-based query in `@Query` and `@Aggregate` annotations use a single parameter or SpEL Expression.

### Examples:
```
 @Query("?0")
 List<Thing> findByComplexFilters(String jsonQuery);

 
 @Query("?#{[0].build()}")
 List<Thing> findByComplexFilters(MyComplexQueryBuilder jsonQueryBuilder);
```

### CODE REVIEW

**Comment by rolag-it** [comment]
I just found the related issue #3874 fixed by @christophstrobl. My fix adds immediate expression result parsing as `org.bson.Document` and a few unit tests. I think that the two approaches can be combined.

**Comment by rolag-it** [comment]
Hello @christophstrobl, I just merged your pull request, please take a look when you can. Thanks

**Comment by christophstrobl** [comment]
Thank you @rolag-it! We'll have a look (might take some time due to the upcoming holiday season though).

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/util/json/ParameterBindingJsonReader.java [hunk 1]
```diff
@@ -124,10 +130,20 @@ public ParameterBindingJsonReader(String json, ParameterBindingContext bindingCo
...
 		this.bindingContext = bindingContext;
 
-		Matcher matcher = PARAMETER_ONLY_BINDING_PATTERN.matcher(json);
+		Matcher matcher = ENTIRE_QUERY_BINDING_PATTERN.matcher(json);
 		if (matcher.find()) {
-			currentValue = bindableValueFor(new JsonToken(JsonTokenType.UNQUOTED_STRING, json)).getValue();
+			BindableValue bindingResult = bindableValueFor(new JsonToken(JsonTokenType.UNQUOTED_STRING, json));
+			try {
+				if (bindingResult.getType().equals(BsonType.DOCUMENT)) {
+					currentValue = Document.parse(bindingResult.getValue().toString());
```
Comment (christophstrobl) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/util/json/ParameterBindingJsonReader.java:138]: This part throws an error when the bound value is an `org.bson.Document` or a `Map`. `StringBasedMongoQueryUnitTests` do not pass at this point.

Comment (rolag-it) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/util/json/ParameterBindingJsonReader.java]: Right, so make more sense to check if expression result is a String

**Comment by christophstrobl** [comment]
thanks @rolag-it!

**Comment by christophstrobl** [comment]
Merged to _main_ development line via (a8a0fb5).

**Pull request closed without merging**